Checked arithmetic by default. (#9465)
ethereum/solidity: docs/control-structures.rst

Prior to Solidity 0.8.0, arithmetic operations would always wrap in case of
under- or overflow leading to widespread use of libraries that would introduce
additional checks.

Since Solidity 0.8.0, all arithmetic operations revert on over- and underflow by default.

To obtain the previous behaviour, an ``unchecked`` block can be used:

::

    // SPDX-License-Identifier: GPL-3.0
    pragma solidity >0.7.99;
    contract C {
        function f(uint a, uint b) pure public returns (uint) {
            // This addition will wrap on underflow.
            unchecked { return a - b; }
        }
        function g(uint a, uint b) pure public returns (uint) {
            // This addition will revert on underflow.
            return a - b;
        }
    }

The call to ``f(2, 3)`` will return ``2**256-1``, while ``g(2, 3)`` will cause
a failing assertion.

The ``unchecked`` block can be used everywhere where a block can be used.

The setting only affects the statements that are syntactically inside the block.
Functions called from within an ``unchecked`` block do not inherit the property.

.. note::
    If you use ``unchecked { _; }`` inside a modifier, this will **not** cause the

>     function to use wrapping arithmetic.

Reviewer: Why do we make it this way? I think this is counterintuitive, unless there is a good reason we want users not to do this.

> An overflow / underflow is the situation that the resulting value of an arithmetic operation,

Reviewer: ```suggestion
An overflow or underflow is the situation where the resulting value of an arithmetic operation,
```

If we want to use `/`, then there shouldn't be a space after it, I think.

> when executed on an unrestricted integer, falls outside the range of the result type.

Reviewer 2: This seems a bit out of place to me here... this should probably be the second paragraph of the section at the very top?

> when executed on an unrestricted integer, falls outside the value range of the type of the result.

Reviewer: ```suggestion
when executed on an unrestricted integer, falls outside the range of the result-type.
```

> The following operators will cause a failing assertion on overflow / underflow

Reviewer: ```suggestion
The following operators will cause a failing assertion on overflow or underflow
```

> and will wrap without error if used inside an unchecked block:

Reviewer: ```suggestion
and will wrap without an error if used inside an unchecked block:
```

``++``, ``--``, ``+``, binary ``-``, unary ``-``, ``*``, ``/``, ``%``, ``**``

``+=``, ``-=``, ``*=``, ``/=``, ``%=``

.. warning::
    It is not possible to disable the check for division by zero
    or modulo by zero using the ``unchecked`` block.

.. note::
    The second statement in ``int x = type(int).min; -x;`` will result in an underflow

>     because the negative range can hold one more value than the positive range.

Reviewer: Is it considered as an underflow?

Reviewer 2: I'd also call it an overflow.

#. If you access an array or an array slice at a too large or negative index (i.e. ``x[i]`` where ``i >= x.length`` or ``i < 0``).
#. If you access a fixed-length ``bytesN`` at a too large or negative index.
#. If you divide or modulo by zero (e.g. ``5 / 0`` or ``23 % 0``).
#. If an arithmetic operation results in under- or overflow outside of an ``unchecked { ... }`` block.

> #. If you shift by a negative amount.

Reviewer: Not relevant to the PR. But is this possible any more? We disallowed using signed types on the RHS of `<<` and `>>`

Author: Fixed: https://github.com/ethereum/solidity/pull/9998
